File: javascript/engines/memory.js
Faye.Engine.Memory = Faye.Class(Faye.Engine.Base, {
  className: 'Engine.Memory',

  initialize: function(options) {
    this._clients   = {};
    this._channels  = {};
    this._messages  = {};
    this._namespace = new Faye.Namespace();
    
    Faye.Engine.Base.prototype.initialize.call(this, options);
  },
  
  createClient: function(callback, scope) {
    var clientId = this._namespace.generate();
    this.debug('Created new client ?', clientId);
    this._clients[clientId] = new Faye.Set();
    this.ping(clientId);
    callback.call(scope, clientId);
  },
  
  destroyClient: function(clientId, callback, scope) {
    var clients = this._clients;
    if (!clients.hasOwnProperty(clientId)) return;
    clients[clientId].forEach(function(channel) {
      this.unsubscribe(clientId, channel);
    }, this);
    this.removeTimeout(clientId);
    delete clients[clientId];
    delete this._messages[clientId];
    this.debug('Destroyed client ?', clientId);
    if (callback) callback.call(scope);
  },
  
  clientExists: function(clientId, callback, scope) {
    callback.call(scope, this._clients.hasOwnProperty(clientId));
  },
  
  ping: function(clientId) {
    var timeout = this._options.timeout;
    if (typeof timeout !== 'number') return;
    this.debug('Ping ?, ?', clientId, timeout);
    this.removeTimeout(clientId);
    this.addTimeout(clientId, 2 * timeout, function() {
      this.destroyClient(clientId);
    }, this);
  },
  
  subscribe: function(clientId, channel, callback, scope) {
    var clients = this._clients, channels = this._channels;
    clients[clientId] = clients[clientId] || new Faye.Set();
    channels[channel] = channels[channel] || new Faye.Set();
    clients[clientId].add(channel);
    channels[channel].add(clientId);
    this.debug('Subscribed client ? to channel ?', clientId, channel);
    if (callback) callback.call(scope, true);
  },
  
  unsubscribe: function(clientId, channel, callback, scope) {
    var clients = this._clients, channels = this._channels;
    if (clients.hasOwnProperty(clientId)) clients[clientId].remove(channel);
    if (channels.hasOwnProperty(channel)) channels[channel].remove(clientId);
    this.debug('Unsubscribed client ? from channel ?', clientId, channel);
    if (callback) callback.call(scope, true);
  },
  
  publish: function(message) {
    this.debug('Publishing message ?', message);

    var channels = Faye.Channel.expand(message.channel),
        messages = this._messages;
    
    Faye.each(channels, function(channel) {
      var clients = this._channels[channel];
      if (!clients) return;
      clients.forEach(function(clientId) {
        this.debug('Queueing for client ?: ?', clientId, message);
        messages[clientId] = messages[clientId] || new Faye.Set();
        messages[clientId].add(message);
        this.emptyQueue(clientId);
      }, this);
    }, this);
  },
  
  emptyQueue: function(clientId) {
    var conn = this.connection(clientId, false),
        messages = this._messages[clientId];
    
    if (!conn || !messages) return;
    delete this._messages[clientId];
    messages.forEach(conn.deliver, conn);
  }
});

Faye.Engine.register('memory', Faye.Engine.Memory);

File: lib/faye/engines/memory.rb
module Faye
  module Engine
    
    class Memory < Base
      def initialize(options)
        @clients   = {}
        @channels  = {}
        @messages  = {}
        @namespace = Namespace.new
        super
      end
      
      def create_client(&callback)
        client_id = @namespace.generate
        debug 'Created new client ?', client_id
        @clients[client_id] = Set.new
        ping(client_id)
        callback.call(client_id)
      end
      
      def destroy_client(client_id, &callback)
        return unless @clients.has_key?(client_id)
        @clients[client_id].each do |channel|
          unsubscribe(client_id, channel)
        end
        remove_timeout(client_id)
        @clients.delete(client_id)
        @messages.delete(client_id)
        debug 'Destroyed client ?', client_id
        callback.call if callback
      end
      
      def client_exists(client_id, &callback)
        callback.call(@clients.has_key?(client_id))
      end
      
      def ping(client_id)
        timeout = @options[:timeout]
        return unless Numeric === timeout
        debug 'Ping ?, ?', client_id, timeout
        remove_timeout(client_id)
        add_timeout(client_id, 2 * timeout) { destroy_client(client_id) }
      end
      
      def subscribe(client_id, channel, &callback)
        @clients[client_id] ||= Set.new
        @channels[channel]  ||= Set.new
        @clients[client_id].add(channel)
        @channels[channel].add(client_id)
        debug 'Subscribed client ? to channel ?', client_id, channel
        callback.call(true) if callback
      end
      
      def unsubscribe(client_id, channel, &callback)
        @clients[client_id].delete(channel) if @clients.has_key?(client_id)
        @channels[channel].delete(client_id) if @channels.has_key?(channel)
        debug 'Unsubscribed client ? from channel ?', client_id, channel
        callback.call(true) if callback
      end
      
      def publish(message)
        debug 'Publishing message ?', message
        channels = Channel.expand(message['channel'])
        channels.each do |channel|
          next unless clients = @channels[channel]
          clients.each do |client_id|
            debug 'Queueing for client ?: ?', client_id, message
            @messages[client_id] ||= Set.new
            @messages[client_id].add(message)
            empty_queue(client_id)
          end
        end
      end
      
    private
      
      def empty_queue(client_id)
        return unless conn = connection(client_id, false) and
               messages = @messages.delete(client_id)
        
        messages.each(&conn.method(:deliver))
      end
    end
    
    register 'memory', Memory
    
  end
end

